### PR TITLE
Initial support for STM32 L5 & U5 devices and minor changes

### LINKED ISSUE #1005: [feature] Support for STM32L5 / STM32L562E-DK board: read_debug32 returns only 0
- Programmer/board type: STM32L562E-DK development board
- Programmer firmware version:
  - Board firmware: 1.0.1
  - STLink version: 3
  - JTAG version: 5
- Operating system and version: Linux 5.6.0
- **Stlink tools version** and/or git commit hash: git hash commit be0157f63329ee48f272fd1a49e709a5ddb6183b
- Stlink commandline tool name: test-usb
- Target chip (and board if applicable): STM32L5

We are having an issue reading debug space on this board. `read_debug32` won't return anything but 0, regardless of what address is specified.

The board is advertising STLink V3SET device (VID=0x0483:PID=0x374f).

`test-usb` utility seems to indicate that some basic functionality works.

Core_ID of 0xbe12477 being returned by `test-usb` utility.  ST documentation seems to indicate that chip_id should be 0x477, which seems to be consistent with existing chip_id enumerations.

With `st-info --probe` test, word one of two word response contains 0x80 or 0x18.  Second word is always 0 which defeats the ability for chip probe to determine the chip_id.

ST documentation indicates that L5 board is a feature enhancement of
L4/L4+ boards but architecture/topology is largely the same.  Coded up
test chip identification structure based on STLINK_CHIPID_STM32_L46X
definition with debug code added to cause
`./src/common:stlink_chip_id()` to a chip_id value of 0x477.

ST documentation indicates that chip has dual bank support with
SRAM1=192K and SRAM2=64K.  SRAM2 is mapped at 0x0a030000 with second
mapping at 0x20030000 to be contiguous with SRAM1.

Flash size data register read for 0x1fff75e0 returns 0 with first word
of read being 0x18.  Additional work would appear to be problematic
without proper functionality of the `read_debug32` function.

Would appreciate any suggestions the group may have for further
investigation.

### LINKED ISSUE #1096: [feature] Support for STM32U5-series
Currently the U5 family is not supported.

- [x] Programmer/board type: STlink/v2, STlink/v2-onboard, STlink/v3
- [x] Operating system: Linux, Mac OS X, Windows (with specific version)
- [x] Stlink tools version and/or git commit hash: v1.6.1
- [x] Stlink commandline tool name: st-info, st-flash, st-util
- [x] Target chip (and optional board): STM32U5-series

### LINKED ISSUE #1194: STM32G031G8: BOOT_LOCK is not possible to change on option bytes address 0x1FFF7870
**Thank you for giving feedback to the stlink project.**

---

**NOTE: In order to offer sufficient and the best possible support, please read /CONTRIBUTING.md and follow the given instructions _before_ submitting a ticket.**

**Bug reports and/or feature requests will be deleted, if they violate our contribution guidelines and if no issue-template is used!** Thank you for your support.

---

- [x] I made serious effort to avoid creating duplicate or nearly similar issue

In order to allow developers to isolate and target your respective issue, please take some time to select the check boxes below and fill out each of the following items appropriate to your specific problem.

- [x] Programmer/board type: [ST-Link V3 mini] (e.g STLINK /V1, /V2, /V2-onboard, /V2-clone, /V3)
- [x] Operating system an version: [Linux] (e.g Linux, macOS, Windows)
- [x] **stlink tools version** and/or git commit hash: [1.7.0-120-gbeffed4-dirty] (e.g v1.6.1/git-d0416149)
- [x] stlink commandline tool name: [st-flash] (e.g `st-info`, `st-flash`, `st-trace`, `st-util`)
- [x] Target chip (and board, if applicable): [ STM32G031G8] (e.g STM32F103C8T6 (NUCLEO-F103RB))

Further we kindly ask you to describe the detected problem as detailed as possible and to add debug output if available, by using the following template:

Commandline output:

./st-flash --debug write output.bin 0x1FFF7870
```
OUTPUT/ERROR of the commandline tool(s)
```
st-flash 1.7.0-120-gbeffed4-dirty
2021-10-18T09:03:32 DEBUG common.c: *** looking up stlink version
2021-10-18T09:03:32 DEBUG common.c: st vid         = 0x0483 (expect 0x0483)
2021-10-18T09:03:32 DEBUG common.c: stlink pid     = 0x374e
2021-10-18T09:03:32 DEBUG common.c: stlink version = 0x3
2021-10-18T09:03:32 DEBUG common.c: jtag version   = 0x3
2021-10-18T09:03:32 DEBUG common.c: swim version   = 0x0
2021-10-18T09:03:32 DEBUG common.c: stlink current mode: mass
2021-10-18T09:03:32 DEBUG usb.c: JTAG/SWD freq set to 0
2021-10-18T09:03:32 DEBUG common.c: stlink current mode: mass
2021-10-18T09:03:32 DEBUG common.c: *** stlink_enter_swd_mode ***
2021-10-18T09:03:32 DEBUG common.c: Loading device parameters....
2021-10-18T09:03:32 DEBUG common.c: *** stlink_core_id ***
2021-10-18T09:03:32 DEBUG common.c: core_id = 0x0bc11477
2021-10-18T09:03:32 DEBUG common.c: *** stlink_read_debug32 0x410cc601 at 0xe000ed00
2021-10-18T09:03:32 DEBUG common.c: *** stlink_read_debug32 0x10016466 at 0x40015800
---------- old ------------
# Chip-ID file for G03x/G04x
#
chip_id 0x466
description G03x/G04x
flash_type  7
flash_size_reg 0x1fff75e0
flash_pagesize 0x800
sram_size 0x2000
bootrom_base 0x1fff0000
bootrom_size 0x2000
option_base 0x1fff7800
option_size 0x4
flags 0

---------- new ------------
# Chip-ID file for G03x/G04x
#
chip_id 0x466
description G03x/G04x
flash_type  7
flash_size_reg 0x0
flash_pagesize 0x800
sram_size 0x2000
bootrom_base 0x1fff0000
bootrom_size 0x2000
option_base 0x1fff7800
option_size 0x4
flags 0

2021-10-18T09:03:32 DEBUG common.c: *** stlink_read_debug32 0xffffffff at 0000000000
2021-10-18T09:03:32 INFO common.c: G03x/G04x: 8 KiB SRAM, 65535 KiB flash in at least 2 KiB pages.
2021-10-18T09:03:32 DEBUG common.c: *** stlink_force_debug_mode ***
2021-10-18T09:03:32 DEBUG common.c: *** stlink_read_debug32 0000000000 at 0x40015808
2021-10-18T09:03:32 DEBUG common.c: *** stlink_write_debug32 0x00001800 to 0x40015808
2021-10-18T09:03:32 DEBUG common.c: *** stlink_status ***
2021-10-18T09:03:33 DEBUG usb.c: core status: 01030003
2021-10-18T09:03:33 DEBUG common.c:   core status: halted
Unknown memory region
2021-10-18T09:03:33 DEBUG common.c: *** stlink_exit_debug_mode ***
2021-10-18T09:03:33 DEBUG common.c: *** stlink_write_debug32 0xa05f0000 to 0xe000edf0
2021-10-18T09:03:33 DEBUG common.c: *** stlink_close ***

Expected/description:

BOOT_LOCK is a flag/bit on address 0x1FFF7870. Due to hard coded option_size of 0x4, it is not possible to change this flag. I get an error "Unknown memory region".

I tried with increased option_size to 0x71, then I get an error "stlink_fwrite_option_bytes() == -1"

### PR DESCRIPTION
**Pull request by Nightwalker-87**
- Initial support for STM32 L5 & U5 devices (Closes #1005) (Closes #1096) (Reference #1247)
- Updated CHANGELOG.md
- Updating windows signal handling to not terminate the application before clean-up.
- Log message improvements for st-flash
- Fix for option bytes on STM32G031G8 (Closes #1194)

### CODE REVIEW

**Comment by Nightwalker-87** [comment]
@Ant-ON Could you give some remarks on this, tell what is still missing and give an overview on what remains to be addressed? This should include the previous attempt of @Evidio from #1247, but should remain as a draft until tested properly with real hardware.

**Comment by Nightwalker-87** [comment]
We need someone with L5 hardware here for final testing.

File: inc/stm32.h [hunk 1]
```diff
@@ -196,10 +205,16 @@ enum stm32_chipids {
...
+#define STM32L5_RCC_DMAEN 0x00000003 // DMA2EN | DMA1EN // RM0438, p. 378
+
 #define STM32H7_RCC_AHB1ENR 0x58024538
 #define STM32H7_RCC_DMAEN 0x00000003 // DMA2EN | DMA1EN
 
 #define STM32WB_RCC_AHB1ENR 0x58000048
 #define STM32WB_RCC_DMAEN 0x00000003 // DMA2EN | DMA1EN
 
+#define STM32L5_PWR_CR1 0x40007000                      // RM0438, p. 93,324
+#define STM32L5_PWR_CR1_VOS 8
```
Comment (Ant-ON) [inline inc/stm32.h:218]: Should be 9

File: src/stlink-lib/common_flash.c [hunk 2]
```diff
@@ -400,22 +423,32 @@ static void unlock_flash(stlink_t *sl) {
...
+    key_reg = get_stm32l0_flash_base(sl) + FLASH_PEKEYR_OFF;
+    flash_key1 = FLASH_L0_PEKEY1;
+    flash_key2 = FLASH_L0_PEKEY2;
+  } else if (sl->flash_type == STM32_FLASH_TYPE_L4) {
+    key_reg = STM32L4_FLASH_KEYR;
+  } else if (sl->flash_type == STM32_FLASH_TYPE_L5_U5) {
+    // Set voltage scaling to range 0 to perform flash operations (RM0438 p. 183)
+    uint32_t mask = (0b11 << STM32L5_PWR_CR1_VOS);
+    uint32_t val;
+    stlink_read_debug32(sl, STM32L5_PWR_CR1, &val);
```
Comment (Ant-ON) [inline src/stlink-lib/common_flash.c:444]: It is better to check the return code, otherwise you can spoil the value of the PWR_CR1 register

**Review by Ant-ON — CHANGES_REQUESTED** [review]

**Comment by Ant-ON** [comment]
@Nightwalker-87 I think L5 support also works like U5. And the problems with the previous PR were due to the fact that the dual bank mode was enabled on the board. To solve the problem of the previous PR, you need to add code below to the `common.c` file after similar code for `STM32_CHIPID_G4_CAT3`.
```c
  if (sl->chip_id == STM32_CHIPID_L5x2) {
    uint32_t flash_optr;
    stlink_read_debug32(sl, STM32L5_FLASH_OPTR, &flash_optr);

    if ((sl->flash_size == 512*1024 &&  (flash_optr & (1 << 22)) != 0) || (sl->flash_size != 512*1024 &&  (flash_optr & (1 << 21)) != 0)) {
      sl->flash_pgsz = 0x800;
    }
  }
```

File: inc/stm32.h (hunk 1 above)
Comment (Nightwalker-87) [reply inc/stm32.h:218]: Fixed. Thanks.

**Comment by Nightwalker-87** [comment]
ok, I entered these lines there.

File: src/stlink-lib/common_flash.c (hunk 2 above)
Comment (Nightwalker-87) [reply src/stlink-lib/common_flash.c:444]: Not sure on how to address this one exactly...

Comment (Nightwalker-87) [reply src/stlink-lib/common_flash.c:444]: Ah, I see... I thought you meant something more complex... Easy then. :+1:

**Comment by slyshykO** [comment]
In this PR there is a removed build for Ubuntu 18. The minimum version remaining is Ubuntu 20.04 with libusb 1.0.23.
So I suggest increasing requirements to libusb according to CI in https://github.com/stlink-org/stlink/blob/f2023_1/src/stlink-lib/libusb_settings.h#L37.

**Review by Ant-ON — APPROVED** [review]

**Comment by Nightwalker-87** [comment]
> In this PR there is a removed build for Ubuntu 18. The minimum version remaining is Ubuntu 20.04 with libusb 1.0.23. So I suggest increasing requirements to libusb according to CI in https://github.com/stlink-org/stlink/blob/f2023_1/src/stlink-lib/libusb_settings.h#L37.

I forgot about that ... :disappointed_relieved:

**Review by slyshykO — APPROVED** [review]

**Pull request merged**